Add aidev-mode recipe

inaimathi opened the pull request:
### Brief summary of what the package does

Exposes a minimal interface for AI-assisted development in any major mode. Specifically,

```
aidev-insert-chat
aidev-refactor-region-with-chat
aidev-refactor-buffer-with-chat
aidev-new-buffer-from-chat
aidev-start-chat
```

supports `ollama`, Claude and ChatGPT as backends with specifiable models.

### Direct link to the package repository

https://github.com/inaimathi/aidev-mode

### Your association with the package

I'm the maintainer and enthusiastic user (although I have to admit, I mostly call `insert-chat` and `refactor-region-with-chat`).

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  I first have to check - are you aware of some of the other packages in this space?  I am writing this from an airport, so excuse my brevity.  In the meantime, quick first pass -

⸺ `aidev-mode.el` with byte-compile using Emacs 30.1:
```
aidev-mode.el:19:11: Error: Cannot open load file: No such file or directory, request
```

⸺ `aidev-mode.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- aidev-mode.el#L202: Could this `point`/`point-max` comparison use `eobp`?
- aidev-mode.el#L190: You may want to consider `(derived-mode-p 'xyz)`
```

`#'load`-check on each file:
```
Loading aidev-mode.el
  aidev-mode.el:Error: Emacs 30.1:
  (file-missing "Cannot open load file" "No such file or directory" "request")
```

⸺ Package and license:
- Prefer equivalent recipe: `(aidev-mode :fetcher github :repo "inaimathi/aidev-mode")`
- Unreachable package URL in aidev-mode.el: `https://github.com/yourusername/aidev-mode`
- aidev-mode.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)

inaimathi commented:
Hello! Pardon the delay.

1. I'm aware of `gptel` (which differs in that it expects to insert
Markdown code into buffers, and consequently seems to support a minimal
number of modes. I've observed `gptel-mode not supported` in
`emacs-lisp-mode`, `common-lisp-mode`, `scad-mode`, `sml-mode`,
`haskell-mode`, `sh-mode` and `python-mode` among others. The goal of
`aidev` is to be a generally available, minimal minor mode for buffer
interaction/generation using LLM bindings. Plausibly, I could achieve that
by requiring `gptel` and relying on it for direct chat interaction and
request infrastructure, but I suspect that `gptel-mode` makes assumptions
about what kind of requests/outputs it provides).
2. Addressed the missing `request` dependency; pardon that, I have it
installed locally for other reasons so it didn't get caught by
`byte-compile-file` when I ran it locally.
3. Addressed the `melpazoid` points
4. Addressed the load-check points
5. Addressed the package and license points




On Sun, Mar 16, 2025 at 3:46 PM Chris Rayner ***@***.***>
wrote:

> Thanks for this. I first have to check - are you aware of some of the
> other packages in this space? I am writing this from an airport, so excuse
> my brevity. In the meantime, quick first pass -
>
> ⸺ aidev-mode.el with byte-compile using Emacs 30.1:
>
> aidev-mode.el:19:11: Error: Cannot open load file: No such file or directory, request
>
> ⸺ aidev-mode.el with melpazoid <https://github.com/riscy/melpazoid>:
>
> - aidev-mode.el#L202: Could this `point`/`point-max` comparison use `eobp`?
> - aidev-mode.el#L190: You may want to consider `(derived-mode-p 'xyz)`
>
> #'load-check on each file:
>
> Loading aidev-mode.el
>   aidev-mode.el:Error: Emacs 30.1:
>   (file-missing "Cannot open load file" "No such file or directory" "request")
>
> ⸺ Package and license:
>
>    - Prefer equivalent recipe: (aidev-mode :fetcher github :repo
>    "inaimathi/aidev-mode")
>    - Unreachable package URL in aidev-mode.el:
>    https://github.com/yourusername/aidev-mode
>    - aidev-mode.el needs *formal* license boilerplate and/or an
>    SPDX-License-Identifier <https://spdx.dev/ids/>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/melpa/melpa/pull/9404#issuecomment-2727615999>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABVKCILTZ4ZHOOYTN3DAPL2UXIHVAVCNFSM6AAAAABY7CV6K6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRXGYYTKOJZHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: riscy]*riscy* left a comment (melpa/melpa#9404)
> <https://github.com/melpa/melpa/pull/9404#issuecomment-2727615999>
>
> Thanks for this. I first have to check - are you aware of some of the
> other packages in this space? I am writing this from an airport, so excuse
> my brevity. In the meantime, quick first pass -
>
> ⸺ aidev-mode.el with byte-compile using Emacs 30.1:
>
> aidev-mode.el:19:11: Error: Cannot open load file: No such file or directory, request
>
> ⸺ aidev-mode.el with melpazoid <https://github.com/riscy/melpazoid>:
>
> - aidev-mode.el#L202: Could this `point`/`point-max` comparison use `eobp`?
> - aidev-mode.el#L190: You may want to consider `(derived-mode-p 'xyz)`
>
> #'load-check on each file:
>
> Loading aidev-mode.el
>   aidev-mode.el:Error: Emacs 30.1:
>   (file-missing "Cannot open load file" "No such file or directory" "request")
>
> ⸺ Package and license:
>
>    - Prefer equivalent recipe: (aidev-mode :fetcher github :repo
>    "inaimathi/aidev-mode")
>    - Unreachable package URL in aidev-mode.el:
>    https://github.com/yourusername/aidev-mode
>    - aidev-mode.el needs *formal* license boilerplate and/or an
>    SPDX-License-Identifier <https://spdx.dev/ids/>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/melpa/melpa/pull/9404#issuecomment-2727615999>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABVKCILTZ4ZHOOYTN3DAPL2UXIHVAVCNFSM6AAAAABY7CV6K6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRXGYYTKOJZHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

riscy commented:
> which differs in that it expects to insert Markdown code into buffers, and consequently seems to support a minimal number of modes

Maybe you've noticed this too, but I have found a lot of LLMs - especially smaller, local models you'd run with ollama - absolutely have a tendency to spit out Markdown-formatted text.  Even "style control" scoring (as [proposed by LMSYS](https://lmsys.org/blog/2024-08-28-style-control/)) considers "number of markdown headers".  For better or worse, Markdown seems to be where the LLMs are at.

> Plausibly, I could achieve that by requiring `gptel` and relying on it for direct chat interaction and request infrastructure, but I suspect that `gptel-mode` makes assumptions about what kind of requests/outputs it provides)

Would you be willing to look into this a bit further?  There are obviously a ton of AI packages being written for Emacs - I've helped index many of them into MELPA.  But I want to make sure we're not imposing choice overload on our users.

karthink commented:
> I'm aware of `gptel` _(which differs in that it expects to insert Markdown code into buffers, and consequently seems to support a minimal number of modes. I've observed `gptel-mode not supported` in `emacs-lisp-mode`, `common-lisp-mode`, `scad-mode`, `sml-mode`, `haskell-mode`, `sh-mode` and `python-mode` among others. The goal of `aidev` is to be a generally available, minimal minor mode for buffer interaction/generation using LLM bindings.

A clarification about gptel:

This has been consistently difficult to communicate, but gptel works in any buffer -- and I do mean _any_ buffer.  For example, it even works _inside_ the minibuffer, in that you can send a query from the minibuffer recursively and insert the response _in_ the minibuffer.  By default, the response is simply inserted after the query, no matter where you are querying from.

gptel users generally default to thinking that LLM interaction requires a minor-mode and it's been difficult to convey that it doesn't, because the interaction is stateless, and you're just replacing or inserting text into buffers.

`gptel-mode` is just a convenience for users who wanted a dedicated chat buffer.  It is not required to use gptel, and I don't use it much myself.

> Would you be willing to look into this a bit further? There are obviously a ton of AI packages being written for Emacs - I've helped index many of them into MELPA. But I want to make sure we're not imposing choice overload on our users.

`aidev-mode` looks good, but for future reference I'd like to mention that there are two comprehensive, unopionated elisp APIs available for writing custom workflows involving LLMs:
- The `llm` package (in GNU ELPA)
- The `gptel-request` API (part of gptel, in MELPA and NonGNU ELPA)

If you're a package author, either of these can handle the LLM interaction for you (including streaming replies) and let you focus on designing the user interface you would like instead.

inaimathi commented:
Hum. Yeah, I assumed `gptel-mode` would be a minor mode that would extend basic AI-interaction keybindings into whatever mode you call it in (like `aidev` does). Noted regarding `gptel-request` (note to future self: https://github.com/karthink/gptel/blob/3d1856f78d242d84ec61f8a854c968b029fdbe34/gptel.el#L2173). 

It looks like it returns a state machine, and is expected to cause effects via callback. I may have more questions about this when I get back to it, but they're probably not appropriate for a MELPA PR discussion thread :p Mind if I poke you about it offline? (If you've got a preferred contact point for it, feel free to send it to my email or reply here).

karthink commented:
> I may have more questions about this when I get back to it, but they're probably not appropriate for a MELPA PR discussion thread :p Mind if I poke you about it offline? (If you've got a preferred contact point for it, feel free to send it to my email or reply here).

Sure!  The [discussions page](https://github.com/karthink/gptel/discussions) on the gptel repo is probably the best place for a, uh, discussion since the information might be useful to other readers.

riscy commented:
Thank you for chiming in @karthink, and thanks again for this @inaimathi.  

I'll merge now, but one last comment I'd make is that `aidev--prepare-system-message` contains a string literal (a prompt) that I feel should be partly customizable to the user.  I suppose they could redefine the whole function, but exposing a defvar might help users tinker.

And, welcome to MELPA. :)
